ultravox/tools/gradio_demo.py

```diff
@@ -1,42 +1,151 @@
+    model_path: str = "fixie-ai/ultravox-v0_3"
+    # "mps", "cuda", or "cpu"
+    device: str = "mps"
+    data_type: str = "float16"
+    default_prompt: str = ""
+    max_new_tokens: int = 200
+    temperature: float = 0
 
 
 def main():
     args = simple_parsing.parse(config_class=DemoConfig)
-    inference = ultravox_infer.UltravoxInference(args.model_path)
+    inference = ultravox_infer.UltravoxInference(
+        args.model_path,
+        device=args.device,
+        data_type=args.data_type,
+        conversation_mode=True,
+    )
 
-    def wrapper(text: str, audio: Tuple[int, np.ndarray]) -> str:
-        sample = datasets.VoiceSample.from_prompt_and_raw(text, audio[1], audio[0])
-        return inference.infer(sample, max_tokens=64).text
+    def add_text(chatbot: gr.Chatbot, text: str) -> gr.Chatbot:
+        return chatbot + [(text, None)]
 
-    inputs = [
-        gr.Textbox(label="Prompt", value=args.default_prompt),
-        gr.Audio(label="Audio", show_download_button=True),
-    ]
-    outputs = [gr.Textbox(label="Output")]
-    examples = [["Transcribe\n<|audio|>", "examples/test16.wav"]]
+    def add_audio(chatbot: gr.Chatbot, audio: str) -> gr.Chatbot:
+        return chatbot + [((audio,), None)]
 
-    gr.Interface(fn=wrapper, inputs=inputs, outputs=outputs, examples=examples).launch(
-        share=True
-    )
+    def process_turn(
+        chatbot: gr.Chatbot,
+        prompt: str,
+        audio: Optional[str] = None,
+        max_new_tokens: int = 200,
+        temperature: float = 0,
+    ):
+        # We want to keep the prompt (mixed audio/text instruction) as is in voice mode, but set it to "" in anticipation of new prompt in text mode.
+        prompt_to_return = prompt
+        if audio:
+            if "<|audio|>" not in prompt:
+                prompt += "<|audio|>"
+            sample = datasets.VoiceSample.from_prompt_and_file(prompt, audio)
+        else:
+            sample = datasets.VoiceSample.from_prompt(prompt)
+            prompt_to_return = ""
+
+        if len(sample.messages) != 1:
+            raise ValueError(
+                f"Expected exactly 1 message in sample but got {len(sample.messages)}"
+            )
+
+        output = inference.infer(
+            sample,
+            max_tokens=max_new_tokens,
+            temperature=temperature,
+        )
+
+        chatbot = chatbot + [(None, output.text)]
+        return chatbot, gr.update(value=prompt_to_return)
+
+    def process_text(chatbot, prompt, max_new_tokens, temperature):
+        return process_turn(
+            chatbot, prompt, max_new_tokens=max_new_tokens, temperature=temperature
+        )
+
+    def process_audio(chatbot, prompt, audio, max_new_tokens, temperature):
+        return process_turn(
+            chatbot,
+            prompt,
+            audio=audio,
+            max_new_tokens=max_new_tokens,
+            temperature=temperature,
+        )
+
+    def gradio_reset():
```

Comment:
This feels fragile, i.e., if the user reloads the page the conversation won't get reset. This makes me think that the actual chat history should be kept here in the chatbot, but we can keep the KV cache in `LocalInference`. This is similar to how the production (OpenAI, etc) services work.

If you think this is too complex to take on here I'd be OK with fixing this in a followup PR.

Reply:
This is fixed by adding one line:
```
demo.load(gradio_reset, [], [chatbot, prompt, audio], queue=False)
```